Asher/swerve

avakimc opened the pull request:
- Added DriveSubsystem and Drive constants
- Added keepHeading logic
- Added Swerve module and module constants
- Added placeholder variables

balghane commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -13,7 +20,106 @@
+    public static final double BACK_LEFT_CHASSIS_ANGULAR_OFFSET = Math.PI;
+    public static final double BACK_RIGHT_CHASSIS_ANGULAR_OFFSET = Math.PI / 2;
+
+    /** SPARK MAX CAN IDs */
+    public static final int FRONT_LEFT_DRIVING_CAN_ID = 11;
+    public static final int REAR_LEFT_DRIVING_CAN_ID = 13;
+    public static final int FRONT_RIGHT_DRIVING_CAN_ID = 15;
+    public static final int REAR_RIGHT_DRIVING_CAN_ID = 17;
+
+    public static final int FRONT_LEFT_TURNING_CAN_ID = 10;
+    public static final int REAR_LEFT_TURNING_CAN_ID = 12;
+    public static final int FRONT_RIGHT_TURNING_CAN_ID = 14;
+    public static final int REAR_RIGHT_TURNING_CAN_ID = 16;
```
These need to be connected with commas for the comment to apply to all, e.g.
```
    public static final int FRONT_LEFT_DRIVING_CAN_ID = 11, REAR_LEFT_DRIVING_CAN_ID = 13;
```

balghane replied:
Same comment applies in some other spots so take a look through. Note this only works when they're all the same type (`int` in this case)

balghane commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -13,7 +20,106 @@
+    public static final double MAX_ANGULAR_SPEED_RAD_PER_SECONDS = 2 * Math.PI; // radians per second
+
+    /**Chassis configuration*/
+    public static final double TRACK_WIDTH = Units.inchesToMeters(26.5);
+    /**Distance between centers of right and left wheels on robot*/
+    public static final double WHEEL_BASE = Units.inchesToMeters(26.5);
```
Append units to the ends of these variable names: `TRACK_WIDTH_METERS`

balghane commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -13,7 +20,106 @@
+    public static final double DRIVING_MOTOR_FREE_SPEED_RPS = NeoMotorConstants.FREE_SPEED_RPM / 60;
+    public static final double WHEEL_DIAMETER_METERS = 0.0762;
+    public static final double WHEEL_CIRCUMFERENCE_METERS = WHEEL_DIAMETER_METERS * Math.PI;
+    /**45 teeth on the wheel's bevel gear, 22 teeth on the first-stage spur gear, 15 teeth on the bevel pinion*/
+    public static final double DRIVING_MOTOR_REDUCTION = (45.0 * 22) / (DRIVING_MOTOR_PINION_TEETH * 15);
```
Can we go through and PLACEHOLDER all the things we're not sure about?
Some of the ones I've spotted:
DRIVING_MOTOR_PINION_TEETH 
DRIVING_MOTOR_REDUCTION 
WHEEL_DIAMETER_METERS (maybe we know this? but should double check and also use `Units.inchesToMeters` to make it clearer)
TRACK_WIDTH 
WHEEL_BASE


If we took the PID constants from last year then we can keep those for now

balghane commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -13,7 +20,106 @@
+    public static final double FREE_SPEED_RPM = 5676;
+  }
+
+  public static final class DriveConstants {
```
Can we name the classes a bit more clearly?
DriveConstants => SwerveDriveConstants
ModuleConstants => SwerveModuleConstants

balghane commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -13,7 +20,106 @@
+    public static final double DRIVING_MIN_OUTPUT = -1;
+    public static final double DRIVING_MAX_OUTPUT = 1;
+
+    public static final double TURNING_P = 1;
+    public static final double TURNING_I = 0;
+    public static final double TURNING_D = 0;
+    public static final double TURNING_FF = 0;
+    public static final double TURNING_MIN_OUTPUT = -1;
+    public static final double TURNING_MAX_OUTPUT = 1;
```
Let's reorganize as we did for the offseason robot. CAN IDs and controllers go in RobotMap, these would go in calibrations, etc.

balghane commented on src/main/java/frc/robot/subsystems/drive/SwerveModule.java:
```diff
@@ -0,0 +1,164 @@
+import frc.robot.Constants.ModuleConstants;
+
+public class SwerveModule {
+  private final CANSparkMax m_drivingSparkMax;
+  private final CANSparkMax m_turningSparkMax;
+
+  private final RelativeEncoder m_drivingEncoder;
+  private final AbsoluteEncoder m_turningEncoder;
+
+  private final SparkMaxPIDController m_drivingPIDController;
+  private final SparkMaxPIDController m_turningPIDController;
+
+  private double m_chassisAngularOffset = 0;
+  private SwerveModuleState m_desiredState = new SwerveModuleState(0.0, new Rotation2d());
```
Fix variable naming to our convention (mostly just remove `m_`

balghane commented on src/main/java/frc/robot/subsystems/drive/SwerveModule.java:
```diff
@@ -0,0 +1,164 @@
+   * MAXSwerve Module built with NEOs, SPARKS MAX, and a Through Bore
+   * Encoder.
+   */
+  public SwerveModule(int drivingCANId, int turningCANId, double chassisAngularOffset) {
```
Can we put in a TODO comments to check all the error codes from these setter functions?

balghane commented on src/main/java/frc/robot/subsystems/drive/SwerveModule.java:
```diff
@@ -0,0 +1,164 @@
+  private final SparkMaxPIDController m_drivingPIDController;
+  private final SparkMaxPIDController m_turningPIDController;
+
+  private double m_chassisAngularOffset = 0;
```
This variable names needs units

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -4,51 +4,197 @@
 public class DriveSubsystem extends SubsystemBase {
-  Pigeon2 pigeon = new Pigeon2(1);
-  CANSparkMax spark = new CANSparkMax(2, MotorType.kBrushless);
+  private double targetHeading;
```
Units? Consider using Rotation2d if it's easier (that has units built in)

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -4,51 +4,197 @@
-  CANSparkMax spark = new CANSparkMax(2, MotorType.kBrushless);
+  private double targetHeading;
+
+  // Create MAXSwerveModules
```
Fix comment

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -4,51 +4,197 @@
+  private double targetHeading;
+
+  // Create MAXSwerveModules
+  private final SwerveModule m_frontLeft = new SwerveModule(
```
Fix variable naming to camelCase

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -4,51 +4,197 @@
+      DriveConstants.REAR_RIGHT_TURNING_CAN_ID,
+      DriveConstants.BACK_RIGHT_CHASSIS_ANGULAR_OFFSET);
+
+  // The gyro sensor
+  private final ADIS16470_IMU m_gyro = new ADIS16470_IMU();
```
As in our offseason code, we need to swap this for a Pigeon2

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -4,51 +4,197 @@
+  /**
+   * Method to drive the robot using joystick info.
+   *
+   * @param xSpeed        Speed of the robot in the x direction (forward).
+   * @param ySpeed        Speed of the robot in the y direction (sideways).
+   * @param rot           Angular rate of the robot.
```
Can we clarify this comments ot reflect that all of these are:
- desired speeds / rates rather than actual speeds and rates
- in [-1, 1]? Rather than in units like m/s or rad/s

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -4,51 +4,197 @@
+  private final SwerveModule m_frontLeft = new SwerveModule(
+      DriveConstants.FRONT_LEFT_DRIVING_CAN_ID,
+      DriveConstants.FRONT_LEFT_TURNING_CAN_ID,
+      DriveConstants.FRONT_LEFT_CHASSIS_ANGULAR_OFFSET);
```
Add a TODO comment that we need to add some angular offsets that are specific to the position of the module. This has to do with how we're setting the angular offset.

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -4,51 +4,197 @@
+    return m_gyro.getRate() * (DriveConstants.GYRO_REVERSED ? -1.0 : 1.0);
+  }
+
+  public void keepHeading(double x, double y, double rot, boolean fieldRelative){
```
Add comments for the function including the function parameters.

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -4,51 +4,197 @@
+    drive(x, y, desiredRotation, fieldRelative);
+  }
+
+  public void choose(double x, double y, double rot, boolean fieldRelative){
```
Add comments here too, and reconsider the function name?

balghane commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -13,7 +19,73 @@
+  public static final int PLACEHOLDER_INT = 1;
+
+  public static final class SwerveModuleConstants {
+    /**The MAXSwerve module can be configured with one of three pinion gears: 12T, 13T, or 14T.
+     * This changes the drive speed of the module (a pinion gear with more teeth will result in a
+     * robot that drives faster).*/
+    public static final int DRIVING_MOTOR_PINION_TEETH = PLACEHOLDER_INT;
```
Probably just remove

balghane commented on src/main/java/frc/robot/Calibrations.java:
```diff
@@ -0,0 +1,29 @@
+import edu.wpi.first.math.controller.PIDController;
+import frc.robot.Constants.SwerveModuleConstants;
+
+public final class Calibrations {
```
Moar comments anywhere you think they add value

balghane commented on src/main/java/frc/robot/Calibrations.java:
```diff
@@ -0,0 +1,29 @@
+    public static final double DRIVING_MIN_OUTPUT = -1;
+    public static final double DRIVING_MAX_OUTPUT = 1;
+    
+    public static final double TURNING_P = PLACEHOLDER_DOUBLE,
+      TURNING_I = PLACEHOLDER_DOUBLE,
+      TURNING_D = PLACEHOLDER_DOUBLE,
```
For this PID, what's the input and output units? Useful to put in a comment. FF probably has different units I think?

balghane commented on src/main/java/frc/robot/RobotMap.java:
```diff
@@ -2,5 +2,19 @@
+    FRONT_RIGHT_TURNING_CAN_ID = 14,
+    REAR_RIGHT_TURNING_CAN_ID = 16;
+  
+  public static final int PIGEON = 1;
```
Is this a CAN ID? Variable name needs clarity

balghane commented on src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java:
```diff
@@ -4,51 +4,215 @@
+  private final WPI_Pigeon2 m_gyro = new WPI_Pigeon2(RobotMap.PIGEON);
+
+  // Odometry class for tracking robot pose
+  SwerveDriveOdometry m_odometry = new SwerveDriveOdometry(
```
fix variable names for camelCase

tails618 committed:
All the swerve code, constants, etc


Added DriveSubsystem and Drive Constants.

Added keepHeading and choose code

- moved stuff from 2022 Calibrations to 2023 Constants if needed

Cleaned up constants/calibrations/etc

Cleaned up variables in SwerveModule

Cleaned up var names and added comments in DriveSubsystem


Removed useless line of code


Renamed gyro to fit naming conventions


Various changes addressing PR comments


Added I/O for driving FF


Fixed error in variable name


Updated SwerveModule Constants
